File: server/internal/transactions/db/categories.go
package transactionsdb

import (
	"context"
	"database/sql"
	"errors"

	"github.com/samber/lo"

	"tallyo/internal/apierror"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"
)

func categoryFromRow(row dbgen.CategoryRow) *model.Category {
	return &model.Category{ID: model.New(model.GlobalIDCategory, row.CatID), Name: row.CatName, Emoji: row.CatEmoji, GroupName: row.GroupName, GroupEmoji: row.GroupEmoji, Kind: model.CategoryKind(row.GroupKind), SortOrder: int32(row.SortOrder), PlaidPFC2Codes: dbutil.SplitCSV(row.PlaidPfc2Codes)}
}

func categoryFromListRow(row dbgen.ListCategoriesRow) *model.Category {
	return categoryFromRow(dbgen.CategoryRow{CatID: row.CatID, CatName: row.CatName, CatEmoji: row.CatEmoji, GroupName: row.GroupName, GroupEmoji: row.GroupEmoji, GroupKind: row.GroupKind, SortOrder: row.SortOrder, GroupID: row.GroupID, PlaidPfc2Codes: row.PlaidPfc2Codes})
}

func (s *Store) Categories(ctx context.Context) ([]*model.Category, error) {
	rows, err := s.q.ListCategories(ctx, dbgen.ListCategoriesParams{})
	return dbutil.MapRows(rows, err, categoryFromListRow)
}

func (s *Store) CategoryGroups(ctx context.Context) ([]*model.CategoryGroup, error) {
	rows, err := s.q.ListCategories(ctx, dbgen.ListCategoriesParams{GroupOrder: true})
	if err != nil {
		return nil, err
	}
	groups := make([]*model.CategoryGroup, 0)
	index := make(map[int64]*model.CategoryGroup)
	for _, row := range rows {
		cat := categoryFromListRow(row)
		groupID := row.GroupID
		group := index[groupID]
		if group == nil {
			group = &model.CategoryGroup{ID: model.New(model.GlobalIDCategoryGroup, groupID), Name: cat.GroupName, Emoji: cat.GroupEmoji, Kind: cat.Kind}
			index[groupID] = group
			groups = append(groups, group)
		}
		group.Categories = append(group.Categories, cat)
	}
	return groups, nil
}

func (s *Store) Category(ctx context.Context, id int64) (*model.Category, error) {
	rows, err := s.q.ListCategories(ctx, dbgen.ListCategoriesParams{CategoryIds: []int64{id}})
	return dbutil.MapFirstRow(rows, err, categoryFromListRow)
}

func (s *Store) CategoryGroupByID(ctx context.Context, id int64) (*model.CategoryGroup, error) {
	row, err := s.q.CategoryGroupByID(ctx, dbgen.CategoryGroupByIDParams{ID: id})
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	catRows, err := s.q.ListCategories(ctx, dbgen.ListCategoriesParams{GroupID: &id})
	if err != nil {
		return nil, err
	}
	group := model.CategoryGroup{ID: model.New(model.GlobalIDCategoryGroup, row.ID), Name: row.Name, Emoji: row.Emoji, Kind: model.CategoryKind(row.Kind)}
	group.Categories = u.Map(catRows, categoryFromListRow)
	return &group, nil
}

func (s *Store) CreateCategoryGroup(ctx context.Context, input model.CreateCategoryGroupInput) (*model.CategoryGroup, error) {
	id, err := s.q.CreateCategoryGroup(ctx, dbgen.CreateCategoryGroupParams{Name: input.Name, Emoji: input.Emoji, Kind: string(input.Kind), SortOrder: 0})
	if err != nil {
		return nil, err
	}
	return s.CategoryGroupByID(ctx, id)
}

func (s *Store) UpdateCategoryGroup(ctx context.Context, input model.UpdateCategoryGroupInput) (*model.CategoryGroup, error) {
	groupID := input.ID.Int64()
	rows, err := s.q.UpdateCategoryGroup(ctx, dbgen.UpdateCategoryGroupParams{Name: input.Name, Emoji: input.Emoji, ID: groupID})
	if err != nil {
		return nil, err
	}
	if rows == 0 {
		return nil, apierror.Publicf("category group %d not found", groupID)
	}
	return s.CategoryGroupByID(ctx, groupID)
}

func (s *Store) DeleteCategoryGroup(ctx context.Context, id int64) (bool, error) {
	if id == UncategorizedCategoryID {
		return false, apierror.Publicf("cannot delete the uncategorized category group")
	}
	var deleted bool
	err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		count, err := q.CountCategoriesByGroup(ctx, dbgen.CountCategoriesByGroupParams{GroupID: id})
		if err != nil {
			return err
		}
		if count > 0 {
			return apierror.Publicf(
				"cannot delete group with %d categor%s; delete all categories first",
				count,
				lo.Ternary(count == 1, "y", "ies"),
			)
		}
		rows, err := q.DeleteCategoryGroup(ctx, dbgen.DeleteCategoryGroupParams{ID: id})
		if err != nil {
			return err
		}
		deleted = rows > 0
		return nil
	})
	if err != nil {
		return false, err
	}
	return deleted, nil
}

func (s *Store) DeleteCategory(ctx context.Context, id int64) (bool, error) {
	if id == UncategorizedCategoryID {
		return false, apierror.Publicf("cannot delete the uncategorized category")
	}
	err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		category, err := q.CategoryInfo(ctx, dbgen.CategoryInfoParams{ID: id})
		if errors.Is(err, sql.ErrNoRows) {
			return apierror.Publicf("category %d not found", id)
		}
		if err != nil {
			return err
		}

		if category.TransactionCount > 0 {
			return apierror.Publicf(
				"cannot delete category with %d transaction%s; reassign them first",
				category.TransactionCount,
				lo.Ternary(category.TransactionCount == 1, "", "s"),
			)
		}
		return q.DeleteCategory(ctx, dbgen.DeleteCategoryParams{ID: id})
	})
	if err != nil {
		return false, err
	}
	return true, nil
}

func (s *Store) ReorderCategories(ctx context.Context, input model.ReorderCategoriesInput) (*model.CategoryGroup, error) {
	groupID := input.GroupID.Int64()
	categoryIDs := model.LocalInt64IDsPtr(input.CategoryIds)
	err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		if err := ensureCategoryGroupExists(ctx, q, groupID); err != nil {
			return err
		}
		for i, catID := range categoryIDs {
			if err := q.UpdateCategorySortOrder(ctx, dbgen.UpdateCategorySortOrderParams{SortOrder: int64(i + 1), ID: catID, GroupID: groupID}); err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return s.CategoryGroupByID(ctx, groupID)
}

File: server-rs/src/transactions/store/categories.rs
use std::collections::HashMap;

use anyhow::{Result, anyhow};
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    apierror::ApiError,
    database::{self, queries},
    schema::CategoryKind,
    transactions::{Category, CategoryGroup},
};

use super::{category_pfc::replace_pfc2_mappings, mapping::category_kind};

pub const UNCATEGORIZED_CATEGORY_ID: i64 = 0;

pub async fn categories(executor: impl Executor<'_, Database = Sqlite>) -> Result<Vec<Category>> {
    queries::list_categories(executor, queries::ListCategoriesParams::default())
        .await
        .map(|rows| rows.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}

pub async fn category_by_id(executor: impl Executor<'_, Database = Sqlite>, id: i64) -> Result<Option<Category>> {
    let ids = [id];
    queries::list_categories(
        executor,
        queries::ListCategoriesParams {
            category_ids: Some(&ids),
            ..Default::default()
        },
    )
    .await
    .map(|rows| rows.into_iter().next().map(Into::into))
    .map_err(Into::into)
}

pub async fn category_groups(executor: impl Executor<'_, Database = Sqlite>) -> Result<Vec<CategoryGroup>> {
    queries::list_categories(
        executor,
        queries::ListCategoriesParams {
            group_order: true,
            ..Default::default()
        },
    )
    .await
    .map(|rows| {
        rows.into_iter()
            .fold((Vec::new(), HashMap::new()), |(mut groups, mut indexes), row| {
                let group_id = row.group_id;
                let category: Category = row.into();
                let index = *indexes.entry(group_id).or_insert_with(|| {
                    groups.push(CategoryGroup {
                        id: group_id,
                        name: category.group_name.clone(),
                        emoji: category.group_emoji.clone(),
                        kind: category.kind,
                        categories: Vec::new(),
                    });
                    groups.len() - 1
                });
                groups[index].categories.push(category);
                (groups, indexes)
            })
            .0
    })
    .map_err(Into::into)
}

pub async fn category_group_by_id(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
) -> Result<Option<CategoryGroup>> {
    let Some(group) = queries::category_group_by_id_opt(executor, queries::CategoryGroupByIdParams { id }).await?
    else {
        return Ok(None);
    };
    Ok(Some(CategoryGroup {
        id: group.id,
        name: group.name,
        emoji: group.emoji,
        kind: category_kind(&group.kind),
        categories: Vec::new(),
    }))
}

pub async fn create_category_group(
    pool: &SqlitePool,
    name: &str,
    emoji: &str,
    kind: CategoryKind,
) -> Result<CategoryGroup> {
    let id = queries::create_category_group(
        pool,
        queries::CreateCategoryGroupParams {
            name,
            emoji,
            kind: &kind.to_string(),
            sort_order: 0,
        },
    )
    .await?
    .id;
    category_group(pool, id).await
}

pub async fn update_category_group(pool: &SqlitePool, id: i64, name: &str, emoji: &str) -> Result<CategoryGroup> {
    if queries::update_category_group(pool, queries::UpdateCategoryGroupParams { name, emoji, id }).await? == 0 {
        return Err(ApiError::bad_input(format!("category group {id} not found")).into());
    }
    category_group(pool, id).await
}

pub async fn delete_category_group(pool: &SqlitePool, id: i64) -> Result<bool> {
    if id == UNCATEGORIZED_CATEGORY_ID {
        return Err(ApiError::bad_input("cannot delete the uncategorized category group").into());
    }
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let count = queries::count_categories_by_group(
                &mut **transaction,
                queries::CountCategoriesByGroupParams { group_id: id },
            )
            .await?
            .count;
            if count > 0 {
                let suffix = if count == 1 { "y" } else { "ies" };
                return Err(ApiError::bad_input(format!(
                    "cannot delete group with {count} categor{suffix}; delete all categories first"
                ))
                .into());
            }
            queries::delete_category_group(&mut **transaction, queries::DeleteCategoryGroupParams { id })
                .await
                .map(|rows| rows > 0)
                .map_err(Into::into)
        })
    })
    .await
}

pub async fn create_category(pool: &SqlitePool, name: &str, emoji: &str, group_id: i64) -> Result<Category> {
    let name = name.to_owned();
    let emoji = emoji.to_owned();
    let id = database::with_tx(pool, |transaction| {
        Box::pin(async move {
            ensure_group(&mut **transaction, group_id).await?;
            queries::create_category(
                &mut **transaction,
                queries::CreateCategoryParams {
                    name: &name,
                    emoji: &emoji,
                    group_id,
                },
            )
            .await
            .map(|row| row.id)
            .map_err(Into::into)
        })
    })
    .await?;
    category(pool, id).await
}

pub async fn update_category(
    pool: &SqlitePool,
    id: i64,
    name: &str,
    emoji: &str,
    group_id: i64,
    plaid_pfc2_codes: Option<&[String]>,
) -> Result<Category> {
    let name = name.to_owned();
    let emoji = emoji.to_owned();
    let plaid_pfc2_codes = plaid_pfc2_codes.map(ToOwned::to_owned);
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            ensure_group(&mut **transaction, group_id).await?;
            if queries::update_category_fields(
                &mut **transaction,
                queries::UpdateCategoryFieldsParams {
                    name: &name,
                    emoji: &emoji,
                    group_id,
                    id,
                },
            )
            .await?
                == 0
            {
                return Err(ApiError::bad_input(format!("category {id} not found")).into());
            }
            if let Some(codes) = plaid_pfc2_codes.as_deref() {
                replace_pfc2_mappings(transaction, id, codes).await?;
            }
            Ok(())
        })
    })
    .await?;
    category(pool, id).await
}

pub async fn delete_category(pool: &SqlitePool, id: i64) -> Result<bool> {
    if id == UNCATEGORIZED_CATEGORY_ID {
        return Err(ApiError::bad_input("cannot delete the uncategorized category").into());
    }
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let Some(category) =
                queries::category_info_opt(&mut **transaction, queries::CategoryInfoParams { id }).await?
            else {
                return Err(ApiError::bad_input(format!("category {id} not found")).into());
            };
            if category.transaction_count > 0 {
                let suffix = if category.transaction_count == 1 { "" } else { "s" };
                return Err(ApiError::bad_input(format!(
                    "cannot delete category with {} transaction{suffix}; reassign them first",
                    category.transaction_count
                ))
                .into());
            }
            queries::delete_category(&mut **transaction, queries::DeleteCategoryParams { id })
                .await
                .map_err(Into::into)
        })
    })
    .await
    .map(|()| true)
}

pub async fn reorder_categories(pool: &SqlitePool, group_id: i64, category_ids: &[i64]) -> Result<CategoryGroup> {
    let category_ids = category_ids.to_vec();
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            ensure_group(&mut **transaction, group_id).await?;
            for (index, id) in category_ids.iter().enumerate() {
                queries::update_category_sort_order(
                    &mut **transaction,
                    queries::UpdateCategorySortOrderParams {
                        sort_order: (index + 1) as i64,
                        id: *id,
                        group_id,
                    },
                )
                .await?;
            }
            Ok(())
        })
    })
    .await?;
    category_group(pool, group_id).await
}

async fn category(pool: &SqlitePool, id: i64) -> Result<Category> {
    category_by_id(pool, id)
        .await?
        .ok_or_else(|| anyhow!("category {id} not found"))
}

async fn category_group(pool: &SqlitePool, id: i64) -> Result<CategoryGroup> {
    let mut group = category_group_by_id(pool, id)
        .await?
        .ok_or_else(|| anyhow!("category group {id} not found"))?;
    group.categories = queries::list_categories(
        pool,
        queries::ListCategoriesParams {
            group_id: Some(id),
            ..Default::default()
        },
    )
    .await?
    .into_iter()
    .map(Into::into)
    .collect();
    Ok(group)
}

async fn ensure_group(executor: impl Executor<'_, Database = Sqlite>, id: i64) -> Result<()> {
    anyhow::ensure!(
        queries::category_group_by_id_opt(executor, queries::CategoryGroupByIdParams { id })
            .await?
            .is_some(),
        ApiError::bad_input(format!("category group {id} not found"))
    );
    Ok(())
}

#[cfg(test)]
pub(super) async fn group_id_of(pool: &SqlitePool, category_id: i64) -> Result<i64> {
    category_groups(pool)
        .await?
        .into_iter()
        .find(|group| group.categories.iter().any(|category| category.id == category_id))
        .map(|group| group.id)
        .ok_or_else(|| anyhow!("category {category_id} has no group"))
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use super::*;
    use crate::{
        money::Cents,
        testutil::transactions::{account, transaction},
    };

    #[tokio::test]
    async fn manages_category_groups_categories_and_mappings() -> Result<()> {
        let pool = crate::database::dbtest::open().await?;
        let group = create_category_group(&pool, "Home", "*", CategoryKind::Expense).await?;
        let first = create_category(&pool, "Test groceries", "*", group.id).await?;
        let second = create_category(&pool, "Test dining", "*", group.id).await?;
        assert_eq!(category_by_id(&pool, first.id).await?, Some(first.clone()));
        assert!(categories(&pool).await?.iter().any(|category| category.id == second.id));
        assert_eq!(
            category_groups(&pool)
                .await?
                .iter()
                .find(|item| item.id == group.id)
                .unwrap()
                .categories
                .len(),
            2
        );

        let seeded = categories(&pool)
            .await?
            .into_iter()
            .find(|category| !category.plaid_pfc2_codes.is_empty())
            .expect("seeded PFC2 mapping");
        let code = seeded.plaid_pfc2_codes[0].clone();
        update_category(
            &pool,
            seeded.id,
            &seeded.name,
            &seeded.emoji,
            group_id_of(&pool, seeded.id).await?,
            Some(&[]),
        )
        .await?;

        let updated_group = update_category_group(&pool, group.id, "Essentials", "#").await?;
        let updated = update_category(
            &pool,
            first.id,
            "Test food",
            "!",
            group.id,
            Some(&[format!(" {code} "), code.clone()]),
        )
        .await?;
        assert_eq!(updated_group.name, "Essentials");
        assert_eq!(updated.plaid_pfc2_codes, [code]);
        assert_eq!(
            reorder_categories(&pool, group.id, &[second.id, first.id])
                .await?
                .categories
                .into_iter()
                .map(|category| category.id)
                .collect::<Vec<_>>(),
            [second.id, first.id]
        );
        assert_eq!(
            delete_category_group(&pool, group.id).await.unwrap_err().to_string(),
            "cannot delete group with 2 categories; delete all categories first"
        );

        let account_id = account(&pool, "category-account").await?;
        let transaction_id = transaction(
            &pool,
            "category-transaction",
            account_id,
            first.id,
            Cents(100),
            "2026-01-01T12:00:00Z".parse()?,
        )
        .await?;
        assert_eq!(
            delete_category(&pool, first.id).await.unwrap_err().to_string(),
            "cannot delete category with 1 transaction; reassign them first"
        );
        super::super::delete_transaction(&pool, transaction_id).await?;
        assert!(delete_category(&pool, first.id).await?);
        assert!(delete_category(&pool, second.id).await?);
        assert!(delete_category_group(&pool, group.id).await?);
        assert!(!delete_category_group(&pool, group.id).await?);
        Ok(())
    }

    #[tokio::test]
    async fn rejects_missing_and_uncategorized_categories() -> Result<()> {
        let pool = crate::database::dbtest::open().await?;
        assert_eq!(
            create_category(&pool, "Missing", "*", 999)
                .await
                .unwrap_err()
                .to_string(),
            "category group 999 not found"
        );
        assert_eq!(
            update_category_group(&pool, 999, "Missing", "*")
                .await
                .unwrap_err()
                .to_string(),
            "category group 999 not found"
        );
        assert_eq!(
            delete_category(&pool, UNCATEGORIZED_CATEGORY_ID)
                .await
                .unwrap_err()
                .to_string(),
            "cannot delete the uncategorized category"
        );
        assert_eq!(
            delete_category_group(&pool, UNCATEGORIZED_CATEGORY_ID)
                .await
                .unwrap_err()
                .to_string(),
            "cannot delete the uncategorized category group"
        );
        Ok(())
    }
}
